src/app/utils/query-parameter-sanitization.ts

```diff
@@ -0,0 +1,396 @@
+    if (!isKeyValuePair(parameter)) {
+      parameter = '<invalid-parameter>';
+    }
+    queryOptionValue += `;${parameter}`;
+  }
+
+  return queryOptionValue;
+}
+
+/**
+ * @param queryOptionValue - The value of the $orderby query option is a
+ * a comma-separated list of expressions which includes a property name, a property path
+ * terminating on a primitive property and/or suffix asc/desc denoting direction of sorting
+ * Examples:
+ * - GET /users?orderby=displayName
+ * - GET /products?$orderby=releasedate asc,rating desc
+ * - GET /categories?$orderby=products/$count
+ */
+function sanitizeOrderByQueryOptionValue(queryOptionValue: string): string {
+  const sortingExpressions = queryOptionValue.split(',');
+
+  sortingExpressions.forEach((expr, index) => {
+    const expressionParts = expr.split(' ').filter(x => x !== ''); // i.e. property name and sort order
+    let propertyName = expressionParts[0].trim();
+    if (!isPropertyName(propertyName)) {
+      propertyName = '<invalid-value>';
+    }
+    let sanitizedExpression = propertyName;
+
+    // Check if sort direction has been included
+    if (expressionParts.length > 1) {
+      let sortDirection = expressionParts[1].trim();
+      if (sortDirection) {
+        if (sortDirection !== 'asc' && sortDirection !== 'desc') {
+          sortDirection = '<unexpected-value>';
+        }
+        sanitizedExpression += ` ${sortDirection}`;
+      }
+    }
+    sortingExpressions[index] = sanitizedExpression;
+  });
+
+  queryOptionValue = sortingExpressions.join(',');
+  return queryOptionValue;
+}
+
+/**
+ * @param queryOptionValue - The value of the $search query option contains the search term and
+ * can also include the target property of your search
+ * Examples:
+ * GET /me/messages?$search="pizza"
+ * GET /me/messages?$search="body:excitement"
+ * GET /groups?$search="description:One" AND ("displayName:Video" OR "displayName:Drive")
+ */
+function sanitizeSearchQueryOptionValue(queryOptionValue: string): string {
+  let sanitizedQueryString: string = '';
+  const searchSegments = queryOptionValue.match(SEARCH_SEGMENT_REGEX);
+  // This means $search value is empty
+  if (searchSegments === null) {
+    return sanitizedQueryString;
+  }
+
+  for (let segment of searchSegments) {
+    segment = segment.trim();
+
+    // No processing needed for logicalOperators operators; append operator to query string.
+    if (LOGICAL_OPERATORS.includes(segment.toLowerCase())) {
+      sanitizedQueryString += ` ${segment}`;
+      continue;
+    }
+
+    // Sanitize segment in the form of "pizza" and "body:excitement"
+    if (QUOTED_TEXT_REGEX.test(segment)) {
+      if (!segment.includes(':')) {
+        sanitizedQueryString += ` <value>`;
+      }
+      else {
+        // Extract property name
+        let propertyName = segment.substring(1, segment.indexOf(':')).trim();
+        if (!isAllAlpha(propertyName)) {
+          propertyName = '<property>';
```

Comment:
Q: What is the reason to redact property names as PII? And it seems to be redacting only in the case of invalid property names. I don't fully understand this if block.

Reply:
It's mainly to handle errors that our users make. I'm making the assumption that if the property name is not  valid, it could contain sensitive data. Just being cautious.

Reply:
fair enough! Thanks for the explanation.